libvast/include/vast/format/arrow.hpp

```diff
@@ -55,4 +56,54 @@ class writer : public format::writer {
   batch_writer_ptr current_batch_writer_;
 };
 
+/// Arrow InputStream API implementation over `std::istream`.
+/// Implemented by adapting `arrow::io::StdinStream adapted to use `std::istream`.
+class arrow_istream_wrapper : public ::arrow::io::InputStream {
+public:
+  explicit arrow_istream_wrapper(std::shared_ptr<std::istream> input);
+  ~arrow_istream_wrapper() override = default;
+
+  ::arrow::Status Close() override;
+
+  bool closed() const override;
+
+  ::arrow::Result<int64_t> Tell() const override;
+
+  ::arrow::Result<int64_t> Read(int64_t nbytes, void* out) override;
+
+  ::arrow::Result<std::shared_ptr<::arrow::Buffer>>
+
+  Read(int64_t nbytes) override;
+
+private:
+  std::shared_ptr<std::istream> input_;
+  int64_t pos_;
+};
+
+class reader final : public vast::format::reader {
+public:
+  /// Constructs an Arrow IPC reader.
+  /// @param options Additional options.
+  /// @param in Input stream containing the IPC data.
+  reader(const caf::settings& options, std::unique_ptr<std::istream> in);
+
+  /// Replace input stream.
+  /// @param in new input stream.
+  void reset(std::unique_ptr<std::istream> in) override;
+
+  caf::error module(vast::module x) override;
+
+  [[nodiscard]] vast::module module() const override;
+
+  [[nodiscard]] const char* name() const override;
+
+protected:
```

Comment:
The `reader` class is marked `final`, so there cannot be a difference between `protected` and `private` visibility here. I would recommend just using the latter to make it a bit less confusing.